### PR TITLE
Fix a coord_sys issue in GapConductance model at recover mode

### PR DESCRIPTION
**Pull request by jiangwen84**
closes ref #15967

### CODE REVIEW

**Comment by dschwen** [comment]
Are you sure that `_current_subdomain_id` has a valid value in `initialSetup()`? What should the value be in an object that exists on multiple subdomains?

**Comment by jiangwen84** [comment]
> Are you sure that `_current_subdomain_id` has a valid value in `initialSetup()`? What should the value be in an object that exists on multiple subdomains?

Yes, I printed out the _current_subdomain_id, and it shows the same number at initial for both standard and recover run. 

`GapConductance` is a material object, and it should have a corresponding subdomain id, correct?

**Comment by dschwen** [comment]
A single material object can exist on multiple subdomains. It is not clear to me how a well defined value can be provided by the framework.

**Comment by jiangwen84** [comment]
Thank a lot for @dschwen 's suggestions :-)  It is ready for another review.

**Comment by jiangwen84** [comment]
@dschwen Do you have any more comments? Thanks!

File: modules/heat_conduction/src/materials/GapConductance.C [hunk 1]
```diff
@@ -214,8 +215,27 @@ GapConductance::GapConductance(const InputParameters & parameters)
 void
 GapConductance::initialSetup()
 {
+  /// set generated from the passed in vector of subdomain names
+  std::set<SubdomainID> subdomain_ids;
+
+  for (auto & name : _subdomain_names)
+    subdomain_ids.insert(_fe_problem.mesh().getSubdomainID(name));
```
Comment (dschwen) [inline modules/heat_conduction/src/materials/GapConductance.C:222]: You can get those with less code by using `BlockRestrictable::blockIDs()`

File: modules/heat_conduction/src/materials/GapConductance.C [hunk 2]
```diff
@@ -214,8 +215,27 @@ GapConductance::GapConductance(const InputParameters & parameters)
...
+
+  // use either block restriction list or list of all subdomains in the mesh
+  const auto & check_subdomains =
+      subdomain_ids.empty() ? _fe_problem.mesh().meshSubdomains() : subdomain_ids;
```
Comment (dschwen) [inline modules/heat_conduction/src/materials/GapConductance.C:226]:
```suggestion
  const auto & check_subdomains = blockRestricted() ? blockIDs() : meshBlockIDs();
```

File: modules/heat_conduction/src/materials/GapConductance.C [hunk 3]
```diff
@@ -149,7 +149,8 @@ GapConductance::GapConductance(const InputParameters & parameters)
     _dof_map(_quadrature ? &_temp_var->sys().dofMap() : NULL),
     _warnings(getParam<bool>("warnings")),
     _p1(declareRestartableData<Point>("cylinder_axis_point_1", Point(0, 1, 0))),
-    _p2(declareRestartableData<Point>("cylinder_axis_point_2", Point(0, 0, 0)))
+    _p2(declareRestartableData<Point>("cylinder_axis_point_2", Point(0, 0, 0))),
+    _subdomain_names(getParam<std::vector<SubdomainName>>("block"))
```
Comment (dschwen) [inline modules/heat_conduction/src/materials/GapConductance.C:153]:
```suggestion
    _p2(declareRestartableData<Point>("cylinder_axis_point_2", Point(0, 0, 0)))
```

File: modules/heat_conduction/include/materials/GapConductance.h [hunk 4]
```diff
@@ -120,4 +120,7 @@ class GapConductance : public Material
 
   Point & _p1;
   Point & _p2;
+
+  /// if this vector is not empty the variables, kernels and materials are restricted to these subdomains
+  std::vector<SubdomainName> _subdomain_names;
```
Comment (dschwen) [inline modules/heat_conduction/include/materials/GapConductance.h:125]:
```suggestion
```

File: modules/heat_conduction/src/materials/GapConductance.C (hunk 1 above)
Comment (jiangwen84) [reply modules/heat_conduction/src/materials/GapConductance.C]: 👍

**Review by dschwen — DISMISSED** [review]

File: modules/heat_conduction/src/materials/GapConductance.C [hunk 5]
```diff
@@ -214,8 +214,20 @@ GapConductance::GapConductance(const InputParameters & parameters)
 void
 GapConductance::initialSetup()
 {
+  /// set generated from the passed in vector of subdomain names
+  const auto & check_subdomains = blockRestricted() ? blockIDs() : meshBlockIDs();
```
Comment (jiangwen84) [inline modules/heat_conduction/src/materials/GapConductance.C:218]: @dschwen I checked one failed test, and it outputs blockRestricted() = 1, but blockIDs() is empty and meshBlockIDs() has two entities. Do you have any thoughts? One thing I am thinking is that the GapConductance seems to be a boundary material, and the boundary info seems to be reasonable. But I am not sure how to get coord_system from boundary information.

Comment (dschwen) [inline modules/heat_conduction/src/materials/GapConductance.C:218]:
```suggestion
  const auto & check_subdomains = blockRestricted() && !blockIDs().empty() ? blockIDs() : meshBlockIDs();
```

**Comment by jiangwen84** [comment]
@dschwen could you merge this if no more comments? Thanks!

**Review by dschwen — APPROVED** [review]

**Pull request merged**